diff --git a/CRM/Aoservicelisting/Form/ProviderApplicationConfirm.php b/CRM/Aoservicelisting/Form/ProviderApplicationConfirm.php
index cbda70a..95e1c6c 100644
--- a/CRM/Aoservicelisting/Form/ProviderApplicationConfirm.php
+++ b/CRM/Aoservicelisting/Form/ProviderApplicationConfirm.php
@@ -73,21 +73,6 @@ class CRM_Aoservicelisting_Form_ProviderApplicationConfirm extends CRM_Aoservice
 
     $this->assign('campValues', $campValues);
 
-    $this->setDefaults($defaults);
-    foreach ($this->_elements as $element) {
-      if (strpos($element->getName(), '[') !== FALSE) {
-         $key = substr($element->getName(), 0, strpos($element->getName(), '['));
-         $arrayKey = substr($element->getName(), strpos($element->getName(), '[') + 1, -1);
-         $element->setValue($defaults[$key][$arrayKey]);
-      }
-      if (!$entryFound) {
-        unset($campValues[$count]);
-      }
-      $count++;
-    }
-    $this->assign('campValues', $campValues);
-
-    $this->setDefaults($defaults);
     $this->freeze();
     $this->addButtons(array(
       array(
diff --git a/CRM/Aoservicelisting/Form/ProviderApplicationForm.php b/CRM/Aoservicelisting/Form/ProviderApplicationForm.php
index cd0fb23..88523bf 100644
--- a/CRM/Aoservicelisting/Form/ProviderApplicationForm.php
+++ b/CRM/Aoservicelisting/Form/ProviderApplicationForm.php
@@ -125,6 +125,9 @@ class CRM_Aoservicelisting_Form_ProviderApplicationForm extends CRM_Aoservicelis
     $this->buildCustom(SERVICELISTING_PROFILE2, 'profile2');
     $this->assign('customDataType', 'Organization');
     $this->assign('customDataSubType', 'service_provider');
+    if (!empty($this->_submitValues['hidden_custom_group_count'][CAMP_CG]) && $this->_submitValues['hidden_custom_group_count'][CAMP_CG] > 1) {
+      $this->assign('cgCount', $this->_submitValues['hidden_custom_group_count'][CAMP_CG]);
+    }
     $this->assign('entityID', $this->organizationId);
     $this->assign('groupID', CAMP_CG);
 
diff --git a/templates/CRM/Aoservicelisting/Form/ProviderApplicationForm.tpl b/templates/CRM/Aoservicelisting/Form/ProviderApplicationForm.tpl
index d1e3b2b..79fe193 100644
--- a/templates/CRM/Aoservicelisting/Form/ProviderApplicationForm.tpl
+++ b/templates/CRM/Aoservicelisting/Form/ProviderApplicationForm.tpl
@@ -136,27 +136,35 @@
   <script type="text/javascript">
     CRM.$(function($) {
       $('.crm-profile legend').hide();
-      CRM.buildCustomData('Organization', 'service_provider', 1);
+      var cgCount = 1;
+      {/literal}
+        {if $cgCount}
+          {literal}
+            var cgCount = {/literal}{$cgCount}{literal};
+          {/literal}
+        {/if}
+      {literal}
+      CRM.buildCustomData('Organization', 'service_provider', 1, 3);
       $('#crm-container.crm-public .label').css('font-size', '16px');
       var serviceProvider = $('[name=listing_type]:checked').val();
       if (serviceProvider == "1") {
         $('.edit-row-organization_name').hide();
         $('.edit-row-organization_email').hide();
-        $('#display_name_public').prop({'checked': true, 'readonly': true});
+        $('*[data-crm-custom="service_provider_details:Display_First_Name_and_Last_Name_in_public_listing"][value="1"]').prop({'checked': true});
+        $('*[data-crm-custom="service_provider_details:Display_First_Name_and_Last_Name_in_public_listing"]').parent('div.content').css('pointer-events', 'none');
       }
       else {
         $('.edit-row-organization_name').show();
         $('.edit-row-organization_email').show();
-        $('#display_name_public').removeAttr('readonly');
       }
       $('[name=listing_type]').on('change', function() {
         if ($(this).val() == "1") {
           $('.edit-row-organization_name').hide();
           $('.edit-row-organization_email').hide();
-          $('#display_name_public').prop({'checked': true, 'readonly': true});
+          $('*[data-crm-custom="service_provider_details:Display_First_Name_and_Last_Name_in_public_listing"][value="1"]').prop({'checked': true});
+          $('*[data-crm-custom="service_provider_details:Display_First_Name_and_Last_Name_in_public_listing"]').parent('div.content').css('pointer-events', 'none');
         }
         else {
-          $('#display_name_public').removeAttr('readonly');
           $('.edit-row-organization_name').show();
           $('.edit-row-organization_email').show();
         }
